src/logger.hpp

```diff
@@ -0,0 +1,130 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <iostream>
+#include "sync/sync_manager.hpp"
+
+#if REALM_ANDROID
+    #include <android/log.h>
+#endif
+
+namespace realm {
+namespace common {
+
+using LoggerFunction = std::function<void(std::string, std::string)>;
+using LoggerLevel = realm::util::Logger::Level;
+using LoggerEntry = std::pair<std::string, std::string>;
+
+class Logger: public realm::util::RootLogger {
+public:
+    void delegate(LoggerFunction&& delegate){
+
+        m_scheduler->set_notify_callback([this, delegate = std::move(delegate)] { 
+
+            std::lock_guard<std::mutex> lock(m_mutex); 
+            while (!m_log_queue.empty()) {
+                auto entry = m_log_queue.front();
+
+                delegate(entry.first, entry.second);  
+                m_log_queue.pop();
+            }
+        });
+    }
+   
+protected:
+    void do_log(LoggerLevel level, std::string message){
+
+        std::lock_guard<std::mutex> lock(m_mutex); 
+        auto LoggerEntry = std::make_pair(map_level[level], message);
+        m_log_queue.push(LoggerEntry);
+
+        #if REALM_ANDROID
+        auto android_log_level = map_android_log_level[level]; 
+        __android_log_print(android_log_level, 
```

Comment:
This does not seem write. Why is logging done here inplace.

Reply:
I think it's a good idea to match the behaviour observed on iOS, where the user can see the sync logs in they stdout there. Also I think it do no harm as its integrated with Android and the user can choose to opt-out by setting the log flag to off.

Reply:
currently there is not logging support for RN. This PR should be adding it. I don't know what you mean by "behaviour observed on iOS" but logging inplace is not what we want to do. 
As I have written above, this PR should implement a way to invoke a JS function which the developer has registered to be called as a callback in RN

Reply:
The idea is that we give the user the option to capture the sync client's log using a callback. If the user decide to use `console.log()`, it will end up in the android device's log.